### PR TITLE
Toasts instead of error-texts

### LINKED ISSUE #416: [FEAT] Coaches: Use toast for errors

### LINKED ISSUE #417: [FEAT] Admins: Use toast for errors

### PR DESCRIPTION
**Pull request by SeppeM8**
fixes #416 
fixes #417 
Also changed for projectspage & state history page

Big part of #329

### CODE REVIEW

**Review by Tiebe-Vercoutter — CHANGES_REQUESTED** [review]
Instead of using try catch blocks you can do toast.promise

**Comment by stijndcl** [comment]
> Instead of using try catch blocks you can do toast.promise
> 
> https://fkhadra.github.io/react-toastify/promise

Great catch, the promise toasts are fancy as hell.

File: frontend/src/components/AdminsComponents/AddAdmin.tsx [hunk 1]
```diff
@@ -42,20 +41,16 @@ export default function AddAdmin(props: { adminAdded: (user: User) => void }) {
...
-            } else {
-                setUsers(users.concat(response.users));
-            }
-
-            setGettingData(false);
-        } catch (exception) {
-            setError("Oops, something went wrong...");
-            setGettingData(false);
+        const response = await toast.promise(getUsersNonAdmin(filter, page), {
+            error: "Failed to receive users",
```
Comment (stijndcl) [inline frontend/src/components/AdminsComponents/AddAdmin.tsx:45]:
```suggestion
            error: "Failed to retrieve users",
```

File: frontend/src/components/AdminsComponents/AddAdmin.tsx [hunk 2]
```diff
@@ -66,54 +61,32 @@ export default function AddAdmin(props: { adminAdded: (user: User) => void }) {
...
+        const success = await toast.promise(addAdmin(user.userId), {
+            pending: "Adding admin",
+            success: "Admin successfully added",
+            error: "Failed to add admin",
+        });
+        if (!success) {
+            toast.error("Failed to add admin", {
+                toastId: "add_admins_failed",
+            });
         }
```
Comment (stijndcl) [inline frontend/src/components/AdminsComponents/AddAdmin.tsx:80]: I don't think this is necessary. The `error` string in `toast.promise()` should already do this automatically?

File: frontend/src/components/AdminsComponents/AddAdmin.tsx [hunk 3]
```diff
@@ -183,11 +155,20 @@ export default function AddAdmin(props: { adminAdded: (user: User) => void }) {
...
+                            variant="primary"
+                            onClick={() => {
+                                if (selected !== undefined) {
+                                    addUserAsAdmin(selected);
+                                }
+                            }}
+                            disabled={selected === undefined}
+                        >
+                            Add admin
+                        </Button>
```
Comment (stijndcl) [inline frontend/src/components/AdminsComponents/AddAdmin.tsx:168]: Perhaps use the common button

File: frontend/src/components/AdminsComponents/RemoveAdmin.tsx [hunk 4]
```diff
@@ -13,30 +13,34 @@ import { ModalContentWarning } from "../Common/styles";
...
-                setError("Something went wrong. Failed to remove admin");
-            }
-        } catch (error) {
-            setError("Something went wrong. Failed to remove admin");
+        if (removed) {
+            props.removeAdmin(props.admin);
+        } else {
+            toast.error("Failed to remove admin", {
+                toastId: "remove_admin_failed",
+            });
```
Comment (stijndcl) [inline frontend/src/components/AdminsComponents/RemoveAdmin.tsx:43]: Same as above, toast.promise() does this automatically

File: frontend/src/components/StudentsComponents/StudentListFilters/StudentListFilters.tsx [hunk 5]
```diff
@@ -75,46 +75,45 @@ export default function StudentListFilters() {
...
+            getStudents(
                 params.editionId!,
                 nameFilter,
                 rolesFilter,
                 alumniFilter,
                 studentCoachVolunteerFilter,
                 requestedPage
-            );
+            ),
+            { error: "Failed to receive students" }
```
Comment (stijndcl) [inline frontend/src/components/StudentsComponents/StudentListFilters/StudentListFilters.tsx:87]:
```suggestion
            { error: "Failed to retrieve students" }
```

File: frontend/src/components/UsersComponents/Coaches/CoachesComponents/AddCoach.tsx [hunk 6]
```diff
@@ -44,20 +44,16 @@ export default function AddCoach(props: { edition: string; refreshCoaches: () =>
...
-            } else {
-                setUsers(users.concat(response.users));
-            }
-
-            setGettingData(false);
-        } catch (exception) {
-            setError("Oops, something went wrong...");
-            setGettingData(false);
+        const response = await toast.promise(getUsersExcludeEdition(props.edition, filter, page), {
+            error: "Failed to receive users",
```
Comment (stijndcl) [inline frontend/src/components/UsersComponents/Coaches/CoachesComponents/AddCoach.tsx:48]:
```suggestion
            error: "Failed to retrieve users",
```

File: frontend/src/components/UsersComponents/Coaches/CoachesComponents/AddCoach.tsx [hunk 7]
```diff
@@ -77,16 +72,17 @@ export default function AddCoach(props: { edition: string; refreshCoaches: () =>
...
+        const success = await toast.promise(addCoachToEdition(user.userId, props.edition), {
+            error: "Failed to add coach",
+            pending: "Adding coach",
+            success: "Coach successfully added",
+        });
+        if (!success) {
+            toast.error("Failed to add coach", {
+                toastId: "add_coach_failed",
+            });
         }
```
Comment (stijndcl) [inline frontend/src/components/UsersComponents/Coaches/CoachesComponents/AddCoach.tsx:84]: Same as above

File: frontend/src/components/UsersComponents/Coaches/CoachesComponents/RemoveCoach.tsx [hunk 8]
```diff
@@ -46,24 +44,29 @@ export default function RemoveCoach(props: {
...
-            }
-        } catch (error) {
-            setError("Something went wrong. Failed to remove coach");
-            setLoading(false);
+        if (removed) {
+            props.removeCoach();
+        } else {
+            toast.error("Failed to remove coach", {
+                toastId: "remove_coach_failed",
+            });
```
Comment (stijndcl) [inline frontend/src/components/UsersComponents/Coaches/CoachesComponents/RemoveCoach.tsx:67]: Same as above

File: frontend/src/components/UsersComponents/Requests/Requests.tsx [hunk 9]
```diff
@@ -67,34 +67,34 @@ export default function Requests(props: { edition: string; refreshCoaches: () =>
         }
 
         setLoading(true);
-        setError("");
-        try {
-            const response = await getRequests(props.edition, searchTerm, page);
+
+        const response = await toast.promise(getRequests(props.edition, searchTerm, page), {
+            error: "Failed to receive requests",
```
Comment (stijndcl) [inline frontend/src/components/UsersComponents/Requests/Requests.tsx:72]:
```suggestion
            error: "Failed to retrieve requests",
```

File: frontend/src/components/UsersComponents/Requests/RequestsComponents/AcceptReject.tsx [hunk 10]
```diff
@@ -14,51 +14,38 @@ export default function AcceptReject(props: {
...
-        } catch (exception) {
-            setError("Failed to accept");
+        const success = await toast.promise(acceptRequest(props.request.requestId), {
+            error: "Failed to accept request",
+            pending: "Accepting request",
+        });
+        if (!success) {
+            toast.error("Failed to accept request", {
+                toastId: "accept_request_failed",
+            });
```
Comment (stijndcl) [inline frontend/src/components/UsersComponents/Requests/RequestsComponents/AcceptReject.tsx:25]: Same as above

File: frontend/src/components/UsersComponents/Requests/RequestsComponents/AcceptReject.tsx [hunk 11]
```diff
@@ -14,51 +14,38 @@ export default function AcceptReject(props: {
...
-            setError("Failed to reject");
+        const success = await toast.promise(rejectRequest(props.request.requestId), {
+            error: "Failed to reject request",
+            pending: "Rejecting request",
+        });
+        if (!success) {
+            toast.error("Failed to reject request", {
+                toastId: "reject_request_failed",
+            });
         }
```
Comment (stijndcl) [inline frontend/src/components/UsersComponents/Requests/RequestsComponents/AcceptReject.tsx:42]: Same as above

File: frontend/src/views/MailOverviewPage/MailOverviewPage.tsx [hunk 12]
```diff
@@ -56,37 +55,37 @@ export default function MailOverviewPage() {
...
-        try {
-            const response = await getMailOverview(editionId, page, searchTerm, filters);
-            if (response.studentEmails.length === 0) {
-                setMoreEmailsAvailable(false);
-            }
-            if (page === 0) {
-                setEmailRows(
+        const response = await toast.promise(
+            getMailOverview(editionId, page, searchTerm, filters),
+            { error: "Failed to receive states" }
```
Comment (stijndcl) [inline frontend/src/views/MailOverviewPage/MailOverviewPage.tsx:60]:
```suggestion
            { error: "Failed to retrieve states" }
```

File: frontend/src/views/UsersPage/UsersPage.tsx [hunk 13]
```diff
@@ -43,34 +44,33 @@ function UsersPage() {
         }
 
         setLoading(true);
-        setError("");
-        try {
-            const response = await getCoaches(params.editionId as string, searchTerm, page);
+        const response = await toast.promise(
+            getCoaches(params.editionId as string, searchTerm, page),
+            { error: "Failed to receive coaches" }
```
Comment (stijndcl) [inline frontend/src/views/UsersPage/UsersPage.tsx:49]:
```suggestion
            { error: "Failed to retrieve coaches" }
```

File: frontend/src/views/projectViews/ProjectsPage/ProjectsPage.tsx [hunk 14]
```diff
@@ -52,37 +52,39 @@ export default function ProjectPage() {
         }
 
         setLoading(true);
-        try {
-            const response = await getProjects(editionId, searchString, ownProjects, page);
-            if (response) {
+        const response = await toast.promise(
+            getProjects(editionId, searchString, ownProjects, page),
+            { error: "Failed to receive projects" }
```
Comment (stijndcl) [inline frontend/src/views/projectViews/ProjectsPage/ProjectsPage.tsx:57]:
```suggestion
            { error: "Failed to retrieve projects" }
```

File: frontend/src/views/projectViews/ProjectsPage/ProjectsPage.tsx [hunk 15]
```diff
@@ -52,37 +52,39 @@ export default function ProjectPage() {
...
             }
-        } catch (exception) {
-            setError("Oops, something went wrong...");
+
+            setPage(page + 1);
+        } else {
+            toast.error("Failed to receive projects", {
+                toastId: "fetch_projects_failed",
+            });
         }
```
Comment (stijndcl) [inline frontend/src/views/projectViews/ProjectsPage/ProjectsPage.tsx:85]: Same as above

**Review by stijndcl — DISMISSED** [review]
Two main things:
- Retrieve is a bit of a nicer word to use than receive, because it fails to fetch them.
- The toast.promise() should already show the `error` message if the request fails. It shouldn't be necessary to check & do this manually again afterwards.

**Review by stijndcl — APPROVED** [review]

**Pull request merged**